glean/rts/factset.h

```diff
@@ -108,6 +108,93 @@ template<typename T, typename By> using FastSetBy =
  *
  */
 class FactSet final : public Define {
+private:
+  class Facts final {
+  public:
+    explicit Facts(Id start) noexcept
+      : starting_id(start) {}
+    Facts(Facts&& other) noexcept
+      : starting_id(other.starting_id)
+      , facts(std::move(other.facts))
+      , fact_memory(other.fact_memory) {}
+    Facts& operator=(Facts&& other) noexcept {
+      starting_id = other.starting_id;
+      facts = std::move(other.facts);
+      fact_memory = other.fact_memory;
+      return *this;
+    }
+
+    bool empty() const {
+      return facts.empty();
+    }
+
+    size_t size() const {
+      return facts.size();
+    }
+
+    Id startingId() const {
+      return starting_id;
+    }
+
+    struct deref {
+      Fact::Ref operator()(const Fact::unique_ptr& p) const {
+        return p->ref();
+      }
+    };
+
+    using const_iterator =
+      boost::transform_iterator<
+        deref,
+        std::vector<Fact::unique_ptr>::const_iterator>;
+
+    const_iterator begin() const {
+      return boost::make_transform_iterator(facts.begin(), deref());
+    }
+
+    const_iterator end() const {
+      return boost::make_transform_iterator(facts.end(), deref());
+    }
+
+    Fact::Ref operator[](size_t i) const {
+      assert (i < facts.size());
+      return facts[i]->ref();
+    }
+
+    void clear() {
+      facts.clear();
+      fact_memory = 0;
```

Comment:
SHould this set StartingId to Invalid?

Reply:
No, why? It should keep its starting id. Invalid isn't a valid starting id, anyway - if you then insert a fact, it would get that id which would be bad.